modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/WiLinkableRowVersionOperations.java

```diff
@@ -0,0 +1,87 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.storage.pagememory.mv;
+
+import static org.apache.ignite.internal.pagememory.util.PageIdUtils.NULL_LINK;
+import static org.apache.ignite.internal.pagememory.util.PartitionlessLinks.writePartitionless;
+import static org.apache.ignite.internal.storage.pagememory.mv.WriteIntentListSupport.removeNodeFromWriteIntentsList;
+import static org.apache.ignite.internal.util.GridUnsafe.pageSize;
+
+import java.util.function.Supplier;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.pagememory.io.DataPageIo;
+import org.apache.ignite.internal.pagememory.io.PageIo;
+import org.apache.ignite.internal.pagememory.util.PageHandler;
+
+class WiLinkableRowVersionOperations implements RowVersionOperations {
+    private final WiLinkableRowVersion rowVersion;
+
+    WiLinkableRowVersionOperations(WiLinkableRowVersion rowVersion) {
+        this.rowVersion = rowVersion;
+    }
+
+    @Override
+    public void removeFromWriteIntentsList(
+            AbstractPageMemoryMvPartitionStorage storage,
+            Supplier<String> operationInfoSupplier
+    ) {
+        assert storage instanceof PersistentPageMemoryMvPartitionStorage;
+
+        removeNodeFromWriteIntentsList(rowVersion, (PersistentPageMemoryMvPartitionStorage) storage, operationInfoSupplier);
+    }
+
+    @Override
+    public long nextWriteIntentLink(long defaultLink) {
+        return rowVersion.nextWriteIntentLink();
+    }
+
+    @Override
+    public long prevWriteIntentLink() {
+        return rowVersion.prevWriteIntentLink();
+    }
+
+    @Override
+    public PageHandler<HybridTimestamp, Object> converterToCommittedVersion() {
+        return UpdateTimestampAndZeroWiLinksHandler.INSTANCE;
+    }
+
+    private static class UpdateTimestampAndZeroWiLinksHandler implements PageHandler<HybridTimestamp, Object> {
+        private static final UpdateTimestampAndZeroWiLinksHandler INSTANCE = new UpdateTimestampAndZeroWiLinksHandler();
+
+        @Override
+        public Object run(
+                int groupId,
+                long pageId,
+                long page,
+                long pageAddr,
+                PageIo io,
+                HybridTimestamp timestamp,
+                int itemId
+        ) {
+            DataPageIo dataIo = (DataPageIo) io;
+
+            int payloadOffset = dataIo.getPayloadOffset(pageAddr, itemId, pageSize(), 0);
+
+            HybridTimestamps.writeTimestampToMemory(pageAddr, payloadOffset + RowVersion.TIMESTAMP_OFFSET, timestamp);
```

Comment:
Do we nullify a `RowId` part?

A wild thought: `RowId` and `Timestamp` may actually use the same physical space, it'll prevent us from wasting 8 bytes. Is this too radical? Or is this already the case?

Reply:
It's a garbage data, why not